### PR TITLE
[BEAM-7389] Add Python snippet for Partition transform

### PR DESCRIPTION
**Pull request by davidcavazos**
Add Python snippet for Partition transform

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by davidcavazos** [comment]
R: @aaltay Can you have a look please?

**Comment by aaltay** [comment]
R: @tvalentyn could you review this PR when you get a chance?

File: sdks/python/apache_beam/examples/snippets/transforms/element_wise/partition_test.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    # [END partitions]
+
+    def partitions_test(actual1, actual2, actual3):
+      assert_that(actual1, equal_to(annuals), label='assert annuals')
+      assert_that(actual2, equal_to(biennials), label='assert biennials')
+      assert_that(actual3, equal_to(perennials), label='assert perennials')
+    self.partitions_test = partitions_test
+
+    # [START train_test]
+    train_dataset = [
```
Comment (tvalentyn) [inline sdks/python/apache_beam/examples/snippets/transforms/element_wise/partition_test.py]: 1. I think we should we relax the requirement which elements fall into `test_dataset` and which into `train_dataset` in this scenario. The transform's semantic does not specify which elements of PCollection will be picked, since PCollections don't have an order of elements, however the test assumes a particular order, which may change when implementation of DirectRunner changes. But we can still verify an expectation that 5 elements were split into 4 + 1 and altogether make up for the whole set of initial elements. 

2. Even though we fix the seed for random, I think the generator implementation may change in a future version of Python, and the sequence of numbers with the same seed might be different on Python 3.7 and Python 3.9. How about we stub out the values for `random()`  via mock.patch?

Comment (davidcavazos) [reply sdks/python/apache_beam/examples/snippets/transforms/element_wise/partition_test.py]: I ended up going with a much more deterministic approach, while also making it work for ratios for arbitrary number of partitions :)

**Comment by davidcavazos** [comment]
Run Portable_Python PreCommit

File: sdks/python/apache_beam/examples/snippets/transforms/element_wise/partition.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        | 'Gardening plants' >> beam.Create([
+            {'icon': '🍓', 'name': 'Strawberry', 'duration': 'perennial'},
+            {'icon': '🥕', 'name': 'Carrot', 'duration': 'biennial'},
+            {'icon': '🍆', 'name': 'Eggplant', 'duration': 'perennial'},
+            {'icon': '🍅', 'name': 'Tomato', 'duration': 'annual'},
+            {'icon': '🥔', 'name': 'Potato', 'duration': 'perennial'},
+        ])
+        | 'Partition' >> beam.Partition(
+            lambda plant, n, train_ratio: int(random.random() > train_ratio),
+            2, train_ratio=0.8,
```
[the remark this answers is not in the document]
Comment (tvalentyn) [reply sdks/python/apache_beam/examples/snippets/transforms/element_wise/partition.py]: Right, but in case of `Map`, `FlatMap`, those `kwargs` are documented, see: https://github.com/apache/beam/blob/05900a43de1f380f124c688cd8d804b5133bda64/sdks/python/apache_beam/transforms/core.py#L1247 

However in Partition docstring we didn't document kwargs, and we could fix that. Filed https://issues.apache.org/jira/browse/BEAM-7671, feel free to take it if you are interested.

File: sdks/python/apache_beam/examples/snippets/transforms/element_wise/partition_test.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        {'icon': '🥕', 'name': 'Carrot', 'duration': 'biennial'},
+    ]
+    perennials = [
+        {'icon': '🍓', 'name': 'Strawberry', 'duration': 'perennial'},
+        {'icon': '🍆', 'name': 'Eggplant', 'duration': 'perennial'},
+        {'icon': '🥔', 'name': 'Potato', 'duration': 'perennial'},
+    ]
+    # [END partitions]
+
+    def partitions_test(actual1, actual2, actual3):
```
[the remark this answers is not in the document]
Comment (davidcavazos) [reply sdks/python/apache_beam/examples/snippets/transforms/element_wise/partition_test.py]: I think I didn't push the changes. Pushing them now, I also addressed the star imports there :)

File: sdks/python/apache_beam/examples/snippets/transforms/element_wise/partition_test.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+#
+
+from __future__ import absolute_import
+from __future__ import print_function
+
+import unittest
+
+import mock
+
+from apache_beam.examples.snippets.transforms.element_wise.partition import *
```
[the remark this answers is not in the document]
Comment (davidcavazos) [reply sdks/python/apache_beam/examples/snippets/transforms/element_wise/partition_test.py]: Got it, changing

**Comment by davidcavazos** [comment]
Run Portable_Python PreCommit

**Comment by tvalentyn** [comment]
@chamikaramj  could you please help merge this once the tests are green? Thanks.

**Comment by chamikaramj** [comment]
Seems like there's a lint error ?

**Comment by davidcavazos** [comment]
This is ready for merge @chamikaramj :)

**Comment by davidcavazos** [comment]
This is ready for merge @chamikaramj :)
